hoomd/mpcd/RejectionVirtualParticleFiller.h

```diff
@@ -0,0 +1,175 @@
+        {
+        GPUArray<Scalar4> tmp_pos(NVirtMax, m_exec_conf);
+        GPUArray<Scalar4> tmp_velTag(NVirtMax, m_exec_conf);
+        m_tmp_pos.swap(tmp_pos);
+        m_tmp_velTag.swap(tmp_velTag);
+        }
+    ArrayHandle<Scalar4> h_tmp_pos(m_tmp_pos, access_location::host, access_mode::overwrite);
+    ArrayHandle<Scalar4> h_tmp_velTag(m_tmp_velTag, access_location::host, access_mode::overwrite);
+
+    // Step 2: Draw the particles and assign velocities simultaneously by using temporary memory
+    unsigned int pidx = 0;
+    unsigned int tag = computeFirstTag(&pidx);
+
+    const Scalar vel_factor = fast::sqrt(m_T->getValue(timestep) / m_mpcd_pdata->getMass());
+
+    for (unsigned int i=0; i < NVirtMax; ++i)
+        {
+        hoomd::RandomGenerator rng(hoomd::RNGIdentifier::RejectionFiller, m_seed, timestep, tag, m_filler_id);
+
+        Scalar3 particle = make_scalar3(hoomd::UniformDistribution<Scalar>(lo.x, hi.x)(rng),
+                                       hoomd::UniformDistribution<Scalar>(lo.y, hi.y)(rng),
+                                       hoomd::UniformDistribution<Scalar>(lo.z, hi.z)(rng));
+
+        if (m_geom->isOutside(particle))
+            {
+            h_tmp_pos.data[pidx] = make_scalar4(particle.x,
+                                                particle.y,
+                                                particle.z,
+                                                __int_as_scalar(m_type));
+
+            hoomd::NormalDistribution<Scalar> gen(vel_factor, 0.0);
+            Scalar3 vel;
+            gen(vel.x, vel.y, rng);
+            vel.z = gen(rng);
+            h_tmp_velTag.data[pidx++] = make_scalar4(vel.x,
+                                                     vel.y,
+                                                     vel.z,
+                                                     __int_as_scalar(tag));
+
+            tag++;
+            }
+        }
+
+    // Allocate memory for the new virtual particles.
+    const unsigned int first_idx = m_mpcd_pdata->addVirtualParticles(pidx);
+
+    ArrayHandle<Scalar4> h_pos(m_mpcd_pdata->getPositions(), access_location::host, access_mode::readwrite);
+    ArrayHandle<Scalar4> h_vel(m_mpcd_pdata->getVelocities(), access_location::host, access_mode::readwrite);
+    ArrayHandle<unsigned int> h_tag(m_mpcd_pdata->getTags(), access_location::host, access_mode::readwrite);
+
+    // copy the temporary data to permanent data
+    for (unsigned int i=0; i < pidx; ++i)
+        {
+        const unsigned int realidx = first_idx + i;
+        // positions
+        h_pos.data[realidx] = h_tmp_pos.data[i];
+        // velocity
+        Scalar4 swp = h_tmp_velTag.data[i];
+        h_vel.data[realidx] = make_scalar4(swp.x,
+                                           swp.y,
+                                           swp.z,
+                                           __int_as_scalar(mpcd::detail::NO_CELL));
+        // tags
+        h_tag.data[realidx] = (unsigned int)swp.w;
```

Comment:
Choose this using the compacted tag list:
```suggestion
        h_tag.data[realidx] = first_tag+i;
```